#12172 Address deprecation warnings within twisted.web

twm opened the pull request:
## Scope and purpose

The goal here is to make `twisted.web` internally warning-free such that use of non-deprecated APIs by user code doesn't produce deprecation warnings.

Many internal warnings were generated by use of the `ErrorPage`/`NoResource`/`ForbiddenResource` classes, which were deprecated in Twisted 22.10.0 by applying `@deprecated` to their `__init__` methods. I've moved the deprecations to the module level so that the private aliases `_UnsafeErrorPage`/`_UnsafeNoResource`/`_UnsafeForbiddenResource` don't generate deprecation warnings. Unfortunately replacing these is not straightforward because many of their callers rely on the HTML interpolation behavior that prompted their deprecation.

- [ ] Address the implication of #12158 for `t.w.xmlrpc`
- [x] Address the implication of #12158 for `t.w.proxy` &rarr; https://github.com/twisted/twisted/issues/12344
- [ ] Fix `test_agent` triggering "Passing method to twisted.internet.ssl.CertificateOptions was deprecated in Twisted 17.1.0."
- [x] Fix `test_cgi` use of `NoResource` and `ErrorPage`
- [ ] Fix `test_tap` warnings
- [x] ~Fix `text_xml` warnings about microdom~ These are import-time warnings so we can ignore them.
- [ ] Fix or ignore https://github.com/twisted/twisted/issues/8227
- [ ] Fix or ignore "This process (pid=31032) is multi-threaded, use of fork() may lead to deadlocks in the child."
- [ ] Prevent future regressions in `twisted.web`.

Fixes #12172.

## Contributor Checklist:

This process applies to *all* pull requests - no matter how small.
Have a look at [our developer documentation](https://docs.twisted.org/en/latest/development/dev-process.html) before submitting your Pull Request.

Below is a non-exhaustive list (as a reminder):

* The title of the PR should describe the changes and starts with the associated issue number, like “#1234 Brief description”.
* A release notes news fragment file was create in src/twisted/newsfragments/ (see: [Release notes fragments docs.](https://docs.twisted.org/en/latest/development/dev-process.html#release-notes-management))
* The automated tests were updated.
* Once all checks are green, request a review by leaving a comment that contains exactly the string `please review`.
  Our bot will trigger the review process, by applying the pending review label
  and requesting a review from the Twisted dev team.

twm committed:
t.w.script: avoid deprecations, add missing test

twm committed:
t.w.resource: deprecate the public aliases

twm committed:
t.w.static: don't use deprecated stuff

twm committed:
t.w.distrib: don't use deprecated stuff

twm committed:
t.w.guard: don't use deprecated stuff

twm committed:
Fix links in the PR template

There's now a redirect that drops the anchor reference from the URL.

twm committed:
Add newsfragment

twm committed:
Address circular imports

twm committed:
Prevent perf regression

Use of deprecatedModuleAttribute in twisted.web.resource impacted
performance of calling `resource.getChildForRequest()`, so import
`getChildForRequest` itself.

itamarst commented on src/twisted/web/resource.py:
```diff
@@ -409,6 +413,26 @@ def __init__(self, message="Sorry, resource is forbidden."):
 NoResource = _UnsafeNoResource
 ForbiddenResource = _UnsafeForbiddenResource
 
+# Explicitly deprecate the public aliases:
```
This is the cause of the performance slowdown. Just, really shouldn't use `deprecatedModuleAttribute` if at all possible. That being said, doing "from twisted.web.resource import whatevs" instead of "from twisted.web import resource" in the right place would hopefully solve much of the slowdown.

glyph replied:
IIRC we hit this once before and that kind of fix addressed it almost completely?

twm replied:
Yes, there were two perf-critical references via the module object (see [5f40096](https://github.com/twisted/twisted/pull/12240/commits/5f4009613c3db8e9348b2a07422b6a95ac518f06) and [7a0bfdc](https://github.com/twisted/twisted/pull/12240/commits/7a0bfdce6761579140d0e11b960716dcf8a23672)).

twm committed:
More perf work

The _IEncodingResource interface is also referenced in a hot path.

twm committed:
Back out script changes

The purpose of moving the deprecation warnings to the module level is to
allow these legacy uses of _UnsafeErrorPage to continue without
generating warnings internally, so I can roll this back.

twm committed:
Finish relocating the deprecations to module level

twm committed:
Add missing test coverage of ResourceScriptWrapper

twm committed:
Switch twcgi to pages
